File: app/channels/little_world_channel.rb
class LittleWorldChannel < ApplicationCable::Channel
  def subscribed
    current_avatar.log_in
    stream_from "little_world_channel"
  end

  def unsubscribed
    current_avatar.log_out
  end

  def speak(data)
    message_text = data["message"].to_s.squish.first(256).gsub("<", "&lt;").presence
    return unless message_text.present?

    avatar = Avatar.find_by(uuid: data["uuid"])
    avatar.update(timestamp: data["timestamp"])
    message = LittleWorldsController.render(partial: 'message', locals: { author: avatar.try(:username), message: message_text, timestamp: avatar.try(:timestamp) })
    ActionCable.server.broadcast "little_world_channel", {uuid: data["uuid"], message: message, timestamp: avatar.try(:timestamp)}
  end

  def ping
    ActionCable.server.broadcast "little_world_channel", {ping: true}
  end

  def pong(data)
    avatar = Avatar.find_by(uuid: data["uuid"])
    return unless avatar.present?
    avatar.broadcast_movement
  end
end

File: app/javascript/src/channels/little_world_channel.js
import consumer from "./consumer"

 setupLittleWorldChannel = function() {
  consumer.subscriptions.create({
    channel: "LittleWorldChannel",
    avatar_uuid: currentPlayer.id
  }, {
    connected: function() {
      littleWorld.connected()
    },
    disconnected: function() {
      littleWorld.disconnected()
    },
    received: function(data) {
      if (data.ping) { return App.little_world.pong() }
      var player = Player.findPlayer(data.uuid)

      if (player == undefined) {
        $(".player[data-id=" + data.uuid + "]").remove()
        return littleWorld.loginPlayer(data)
      }

      player.reactToData(data)
    },
    speak: function(msg) {
      return this.perform("speak", {
        uuid: currentPlayer.id,
        message: msg,
        timestamp: nowStamp()
      })
    },
    ping: function() {
      return this.perform("ping")
    },
    pong: function() {
      return this.perform("pong", {
        uuid: currentPlayer.id
      })
    }
  });
}
